power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/dataset.hpp

```diff
@@ -36,16 +38,125 @@ static_assert(is_indptr_mutable_v<writable_dataset_t>);
+                        AttributeType const& attribute_ref = meta_attribute.template get_attribute<AttributeType const>(
+                            reinterpret_cast<RawDataConstPtr>(&value));
+                        *buffer_ptr = attribute_ref;
+                    });
+            }
+            return *this;
+        }
+        operator value_type() const { return get(); }
+        value_type get() const {
+            value_type result{};
+            for (auto const& attribute_buffer : attribute_buffers_) {
+                assert(attribute_buffer.meta_attribute != nullptr);
+                auto const& meta_attribute = *attribute_buffer.meta_attribute;
+                ctype_func_selector(
+                    meta_attribute.ctype, [&result, &attribute_buffer, &meta_attribute, this]<typename AttributeType> {
+                        AttributeType const* buffer_ptr =
+                            reinterpret_cast<AttributeType const*>(attribute_buffer.data) + idx_;
+                        AttributeType& attribute_ref =
+                            meta_attribute.template get_attribute<AttributeType>(reinterpret_cast<RawDataPtr>(&result));
+                        attribute_ref = *buffer_ptr;
+                    });
+            }
+            return result;
+        }
+
+      private:
+        friend class ColumnarAttributeRange;
+        friend class iterator;
+
+        Idx idx_{};
+        std::span<AttributeBuffer<Data> const> attribute_buffers_{};
+    };
+
+    class iterator : public boost::iterator_facade<iterator, T, boost::random_access_traversal_tag, Proxy, Idx> {
+      public:
+        using value_type = Proxy;
+
+        iterator() = default;
+        iterator(Idx idx, std::span<AttributeBuffer<Data> const> attribute_buffers)
+            : current_{idx, attribute_buffers} {}
+
+      private:
+        friend class boost::iterator_core_access;
+
+        constexpr auto dereference() const { return current_; }
+        constexpr auto equal(iterator const& other) const { return current_.idx_ == other.current_.idx_; }
+        constexpr auto distance_to(iterator const& other) const { return other.current_.idx_ - current_.idx_; }
+        constexpr void increment() { ++current_.idx_; }
+        constexpr void decrement() { --current_.idx_; }
+        constexpr void advance(Idx n) { current_.idx_ += n; }
+
+        Proxy current_;
+    };
+
+    ColumnarAttributeRange() = default;
+    ColumnarAttributeRange(Idx size, std::vector<AttributeBuffer<Data>> attribute_buffers)
+        : size_{size}, attribute_buffers_{std::move(attribute_buffers)} {}
+    ColumnarAttributeRange(ColumnarAttributeRange::iterator begin, ColumnarAttributeRange::iterator end)
+        : size_{std::distance(begin, end)},
+          attribute_buffers_{begin->attribute_buffers_.begin(), begin->attribute_buffers_.end()} {
+        assert(begin + std::distance(begin, end) == end);
+    }
```

Comment:
i have a feeling there's still a bug in this: let's say you have a range `[0, 1, 2, 3, 4]` and you take sub-range `[1, 2, 3]` (like we do in `get_columnar_buffer_span_impl`), then you take:

* `auto const start = total_range.begin() + 1` ==> `iterator{Proxy{.idx_ = 1, .attribute_buffers_ = total_range.attribute_buffers}}`
* `auto const stop = total_range.begin() + 4` ==> `iterator{Proxy{.idx_ = 4, .attribute_buffers_ = total_range.attribute_buffers}}`
* create new `result = ColumnarAttributeRange(start, stop)` ==> `ColumnarAttributeRange{.size = 4 - 1 = 3, .attribute_buffers_ = {begin->attribute_buffers_.begin(), begin->attribute_buffers_.end()} = std::vector(total_range.attribute_buffers_.begin(), total_range.attribute_buffers_.end()) = copy of total_range.attribute_buffers}`
* meaning `result.attribute_buffers_ == total_range.attribute_buffers_`, so that
* `result[0].get() == total_range[0].get()`
* this is incorrect, because we need `result[1].get() == total_range[0].get()`

(please add a test to verify that this is indeed true)

we need to add an `Idx start_` to this `ColumnarAttributeRange` and below set `iterator get(Idx idx) const { return iterator{start_ + idx, attribute_buffers_}; }`

NOTE: this wasn't an issue before because we could offset the data buffers for all attributes individually before when doing the range copy. Now, we can't anymore. That this also means that now, the entire `ColumnarAttributeRange` itself can also just directly contain an `std::span<AttributeBuffer const>` instead of an owned `std::vector<AttributeBuffer>`

Reply:
I get your point, but I think you made a typo: What we should be getting is `result[0].get() == total_range[1].get()`, shouldn't we? I will add a test to first check this and then implement the fix.

Reply:
Whoops yes you're right. Covariance vs contravariance traumas all over again

Reply:
The bug is indeed present. A test has been added to confirm it.

Reply:
Test passes now with new implementation after introducing `start_` for the `ColumnarAttributeRange` constructor with `iterator`.